loader/windows/icd_windows_dxgk.c

```diff
@@ -118,9 +118,15 @@ bool khrIcdOsVendorsEnumerateDXGK(void)
             Status = D3DKMTQueryAdapterInfo(&queryAdapterInfo);
             if (!NT_SUCCESS(Status))
             {
-                // Continue trying to get as much info on each adapter as possible.
-                // It's too late to return FALSE and claim WDDM2_4 enumeration is not available here.
-                continue;
+                // Try a different value type.  Some vendors write the key as a multi-string type.
+                queryArgs.ValueType = REG_MULTI_SZ;
+                Status = D3DKMTQueryAdapterInfo(&queryAdapterInfo);
```

Comment:
This seems reasonable.  A few administrative topics:

1. I think we should bump the ICD loader version number with this change, so ICD loaders with this fix are considered newer and overwrite ICD loaders without the change.
1. Do we want to consider this a workaround for drivers/installers that do not follow the cl_khr_icd spec, or do we want to update the spec to say that these keys may either be REG_SZ or REG_MULTI_SZ?
1. Do we want to apply a similar change to the HKR codepath?  I believe it is currently only checking for REG_SZ.
1. Regardless, I'd suggest adding some KHR_ICD_TRACE messages to trace when the ICD loader switches from a REG_SZ key to a REG_MULTI_SZ key.

Thanks!

Reply:
Hi,

I've incorporated points 1, 3 & 4 into a second commit.  I have tested the DXGK and HKR code paths briefly on my local system and both enumerate all platforms as expected with these changes.

Regarding point 2.  I would suggest we consider this a workaround.  Comments in the closed pull request #85 suggest AMD will eventually abide by the specification but there are drivers in the field necessitating this workaround.

Thanks for the code review!
John A.

Reply:
Re point 2 (workaround), I'm quoting Ben's response on this matter:
"drivers should be using REG_SZ, see Section 2.4 in the [cl_khr_icd spec](https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_Ext.html#cl_khr_icd-opencl). This change is a workaround for older drivers that did not properly implement the spec and used REG_MULTI_SZ instead, which otherwise do not enumerate properly when used with newer ICD loaders."

Something is not clear to me.
The ICD Loader is (usually) shipped and coupled with the accompanying driver (say Graphics Driver or independent OpenCL Runtime/Platform implementation), so whenever this workaround is applied it means that you shipped a new driver bundle.
Why not fix the root cause and adapt the INF file that comes with the Graphics Driver to use REG_MULTI_SZ?
You're already shipping a new ICD Loader, why not ship a "fixed"/conforming driver while you're at it?

In other words, under what environments does this bug reproduce? hybrid? (e.g. Intel CPU + AMD GPU)

Reply:
This is reproducible on a system with Intel CPU Opencl, Intel GPU Opencl and AMD GPU Opencl (Vega-m), all latest drivers. The AMD loader enumerates only the AMD platform devices, the Intel loader enumerates only the Intel platform devices. I would suggest it is correct to be defensive and enumerate all platforms, including those that have deviated from the spec.

Reply:
Then what good is the spec for :)? is it a recommendation for vendors or something to adhere to?
It's OK to put workaround for the greater good, or at least temporarily, but IMHO it should be discouraged.

Reply:
On second thought, we are required (or at least expected) to be backwards compatible in some sense. So if we changed from REG_SZ to REG_MULTI_SZ at some point in time, it makes sense to accept both as the loader should be able to enumerate older conforming platform implementations.